system/lib/wasmfs/wasmfs.cpp

```diff
@@ -8,60 +8,258 @@
+  }
+};
+
+class StderrFile : public File {
+public:
   // TODO: May not want to proxy stderr (fd == 2) to the main thread.
   // This will not show in HTML - a console.warn in a worker is suffficient.
   // This would be a change from the current FS.
-  if (fd == 1 || fd == 2) {
-    __wasi_size_t num = 0;
-    for (size_t i = 0; i < iovs_len; i++) {
-      const uint8_t* buf = iovs[i].buf;
-      __wasi_size_t len = iovs[i].buf_len;
-      for (__wasi_size_t j = 0; j < len; j++) {
-        uint8_t current = buf[j];
-        if (current == 0 || current == 10) {
-          fd_write_stdstream_buffer.push_back('\0'); // for null-terminated C strings
-          if (fd == 1) {
-            emscripten_console_log(&fd_write_stdstream_buffer[0]);
-          } else if (fd == 2) {
-            emscripten_console_error(&fd_write_stdstream_buffer[0]);
-          }
-          fd_write_stdstream_buffer.clear();
-        } else {
-          fd_write_stdstream_buffer.push_back(current);
-        }
+  __wasi_errno_t write(
+    const __wasi_ciovec_t* iovs, size_t iovs_len, __wasi_size_t* nwritten) override {
+    return write_buffer(iovs, iovs_len, nwritten, &emscripten_console_error);
+  }
+};
+
+class OpenFile {
+  std::shared_ptr<File> file_node;
+  uint32_t _offset;
+  std::mutex openfile_mutex;
+
+public:
+  OpenFile(uint32_t offset, std::shared_ptr<File> _file_node) {
+    this->_offset = offset;
+    this->file_node = _file_node;
+  }
+
+  std::shared_ptr<File> get_file_node() { return file_node; }
+};
+
+class FileTable {
+  std::vector<std::shared_ptr<OpenFile>> handles;
+  std::mutex ft_mutex;
+
+public:
+  std::vector<std::shared_ptr<OpenFile>> get_file_handles() { return handles; }
+  std::shared_ptr<OpenFile> get_open_file(__wasi_fd_t fd) {
+    // Check if the file descriptor is invalid.
+    const std::lock_guard<std::mutex> lock(ft_mutex);
+    if (fd >= handles.size() || fd < 0)
+      return nullptr;
+
+    return handles[fd];
+  }
+
+  __wasi_fd_t add_open_file(std::shared_ptr<OpenFile> ptr) {
+    const std::lock_guard<std::mutex> lock(ft_mutex);
+    for (__wasi_fd_t i = 0; i < handles.size(); i++) {
+      if (!handles[i]) {
+        // Free open file entry.
+        handles[i] = ptr;
+        return i;
       }
-      num += len;
     }
-    *nwritten = num;
+
+    // Could not find an empty open file table entry.
+    handles.push_back(ptr);
+
+    return handles.size() - 1;
   }
-  return 0;
+
+  int remove_open_file(__wasi_fd_t fd) {
+    const std::lock_guard<std::mutex> lock(ft_mutex);
+    // Check if the file descriptor is invalid.
+    if (fd >= handles.size() || fd < 0)
+      return -1;
+
+    handles[fd] = nullptr;
+    return 0;
+  }
+
+  int set_open_file(__wasi_fd_t fd, std::shared_ptr<OpenFile> ptr) {
+    const std::lock_guard<std::mutex> lock(ft_mutex);
+    if (fd < 0)
+      return -1;
+
+    if (fd >= handles.size()) {
+      handles.resize(fd);
+      handles.insert(handles.begin() + fd, ptr);
```

Comment:
Off-by-one error here:
```suggestion
      handles.resize(fd + 1);
      handles[fd] = ptr;
```

Reply:
Oh yeah there isn't a 1 off error here. I can change it make it clearer.